### ISSUE TITLE
Extension not starting

### ISSUE THREAD

**Issue by PvdBerg1998** [dgriffen/rls-vs2017#7]
I've installed the two extensions into the latest VS preview, then opened a rust project. I've also tried manually installing the RLS (currently only available via the beta channel).

> The extension will start when you open a Rust file. You'll be prompted to install the RLS. Once installed, the RLS should start building your project.

.. but it doesn't.

**Comment by dgriffen** [comment]
Do you at least get the prompts to install the RLS? If you do and the RKS doesn't install then the extension is at least starting. If you do get those prompts then it sounds like the problem is that the RLS is currently unavailable on nightly. If you wish to use a different toolchain with this extension you need to set that option in `Tools > Options > Rust`.

**Comment by PvdBerg1998** [comment]
I've set the extension to use the beta  channel. Any idea why it still exits?

**Comment by dgriffen** [comment]
So just to confirm,
- The `Tools>Options>Rust` Toolchain to use setting is set to `beta`.
- You get the installation prompt at the beginning or the necessary components are already installed
    - rls-preview
    - rust-analysis
    - rust-src

After this the extension still exits, is that correct? If so, try restarting VS first and making sure the setting stuck since I am unable to reproduce on any of my systems.

**Comment by dgriffen** [comment]
A few more questions:
- Does your beta build call the RLS component `rls-preview` or `rls`?
- How are you confirming that the extension exits?

**Comment by PvdBerg1998** [comment]
- Setting is correct (beta channel)
- Components are indeed installed
- I've installed `rls-preview` on the beta toolchain
- I do not, I just assumed so since nothing happens.

**Comment by dgriffen** [comment]
If the components are already installed my extension won't pop up any notifications. It's main job is to launch the rls process and pass that on to the LSP extension. You can try looking in task manager to see if there is an `rls.exe` process running (If you are on windows 10 you may need to expand the process tree for visual studio). Tonight I'll add some logging to the extension and send out an update. I'll update this issue with instructions on how to read the logs once the update is out.

It is also possible that the RLS is crashing on your project. If you are working with a particularly large project create a new library via cargo and see what happens then.

**Comment by PvdBerg1998** [comment]
Was using a helloworld project so that shouldn't be the problem. The process is not being started though.

**Comment by dgriffen** [comment]
Its ugly, but I've attached a debug version of the extension [RustLanguageExtension.zip](https://github.com/dgriffen/rls-vs2017/files/1564202/RustLanguageExtension.zip) unzip that and install the .vsix file. You should now get a lot of infobars when you open a rust file. Please tell me which ones you see and which is the last one.

**Comment by PvdBerg1998** [comment]
Could you push the changes (to a separate branch)? I'm trying to build the extension myself.

**Comment by dgriffen** [comment]
I could, I just figured that just providing a vsix was simpler than having you build the extension. Allthough if you build it yourself you can debug the OnLoadedAsync method to see whats going on.

**Comment by dgriffen** [comment]
branch https://github.com/dgriffen/rls-vs2017/tree/super-infobar-debug has the changes

**Comment by PvdBerg1998** [comment]
I only see "extension has begun launching" (or something similar) pop up 🤔

**Comment by dgriffen** [comment]
I assume you pressed continue? Well that means that it is detecting all the components as installed, but it still isn't starting. This was something I didn't predict, let me add a few more info bars.

**Comment by PvdBerg1998** [comment]
I'm trying to debug the extension but it says: "the breakpoint will currently not be hit, no symbols have been loaded for this document".

**Comment by dgriffen** [comment]
It won't bind until you actually load the extension via opening a rust file.

**Comment by PvdBerg1998** [comment]
It seems I can't debug further than the `Rustup.HasToolchain` call, but I'm not experienced with debugging c# vs extensions. I can build it though, you don't need to bother uploading packages :)

**Comment by PvdBerg1998** [comment]
Still only the first popup. Trying a popup at the end of OnLoadedAsync

**Comment by dgriffen** [comment]
Can you try debugging into `HasToolchain`? I think the problem might be an await command that is never exiting, since you couldn't debug past it.

**Comment by PvdBerg1998** [comment]
That's what I'm thinking as well. I'm never getting past the await. The command should be working though.

**Comment by PvdBerg1998** [comment]
`RunCommand` isn't even reached.

**Comment by dgriffen** [comment]
You are pressing continue on the infobars right? cause those block execution. If so, wat. The only call that could be blocking between the first infobar and RunCommand would be OptionsModel.Toolchain.

**Comment by PvdBerg1998** [comment]
I am. Maybe VS's await debugger is broken in the preview? It generates a state machine so I could understand if debugging such a thing would break. I'll try adding print statements somewhere in between.

**Comment by dgriffen** [comment]
The await debugger is pretty rock solid, I haven't had issues with it when I'm debugging. There must be something really screwy going on. I'm guessing something is redirecting a dll my extension uses. (Either via GAC or a binding redirect). You could try removing the `()` tuples from `Rustup.rs` and replacing them with the `Tuple` class.

**Comment by PvdBerg1998** [comment]
```
on loaded sync entered
got toolchain, awaiting rustup.hastoolchain
Exception thrown: 'System.IO.FileNotFoundException' in RustLanguageExtension.dll
```

**Comment by dgriffen** [comment]
Do you not have rustup on your path?

**Comment by PvdBerg1998** [comment]
```
C:\Windows\System32>rustup --version
rustup 1.7.0 (813f7b7a8 2017-10-30)
```

**Comment by PvdBerg1998** [comment]
NB I've added a debug println inside HasToolchain and that didn't print anything - so maybe the Rustup.cs didn't get compiled?

**Comment by dgriffen** [comment]
Rustup.cs must get compiled, you'd have much different errors if it wasn't included in the output binary. How is your rustup path set? It looks like you were running cmd.exe as admin since it doesn't default to the system32 directory otherwise. Perharps its only on the admin path?

**Comment by PvdBerg1998** [comment]
My rustup path is set in my user path, and my user cmd defaults to sys32 for some reason. I've pushed my extra prints if you need them https://github.com/PvdBerg1998/rls-vs2017/commit/9b5c689f7615fd18368ea02cdee1b555fadba70d .

**Comment by dgriffen** [comment]
Did that FileNotFoundException have a stack trace or anything? Because at this point my guess is that you are missing some dll on the system. Since its happening when you hit Rustup.rs it could be System.ValueTuple, which is meant to be shipped with VS. Since I can't repro on your system I pushed an update to the infobar branch that remove the valuetuples.

**Comment by PvdBerg1998** [comment]
This fixes the problem. (And it didn't have a stack trace btw.)

I do get a lot of `System.Windows.Data Error: 5 : Value produced by BindingExpression is not valid for target property.; Value='<null>' BindingExpression:Path=IconAutomationText; DataItem='CompletionView' (HashCode=16558849); target element is 'CrispImage' (Name=''); target property is 'Name' (type 'String')` spam though, but that might be a different issue.

**Comment by dgriffen** [comment]
System.ValueTuple really should be included with a VS installation, it sounds like yours got corrupted somehow. Try repairing it to see if that fixes it.

**Comment by DustinByfuglien** [comment]
Hello.
I install LSP, extension but nothing happens when I open rust file.
No error or other messages is  provided.
I'm confused...

Win7x64

**Comment by dgriffen** [comment]
Did you open an individual file or an entire folder? The RLS and the LSP extension don't work with individual files.

**Comment by DustinByfuglien** [comment]
I open a folder. Than open a main.rs file.

**Comment by dgriffen** [comment]
Hopefully its not a Win7 specific issue as I don't have any machine to test that with. It may be the same issue as PvdBerg so repairing your VS installation may be the way to go. I could remove ValueTuple to get it working for you immediately but your VS install would still be unstable since ValueTuple is used throughout VS.

**Comment by DustinByfuglien** [comment]
After VS repairing and opening prjoect folder a toolbar with rls Install button appears.
But after clicking on it a error rises in that tab: 

> could not install one of the required rls components

**Comment by dgriffen** [comment]
That will likely be an issue with the RLS not being available in the current nightly. It's been a major issue lately and is quite annoying. You can select your toolchain in the Tools>Options>Rust menu

**Comment by DustinByfuglien** [comment]
Thanks. I'll wait a while with nightly.

**Comment by KiillThemAll** [comment]
Hello!
Have this issue, only see "extension has begun launching"
Tryed to repair VS - useless
Have any advices?

**Comment by dgriffen** [comment]
What do you mean by seeing "extension has begun launching"? That's not a prompt that VS would normally display to you.

**Comment by KiillThemAll** [comment]
Tryed your extension with more info bars.
RLS manually installed
Windows 10

**Comment by dgriffen** [comment]
Yeah, sounds like the `System.ValueTuple` issue again. I'll talk to people within VS to figure out what could be going on. VS includes `System.ValueTuple` so there should be no issue using it. Sorry for the inconvenience. If I'm not able to figure out the issue by the end of the week I'll just remove tuple stuff.

**Comment by dgriffen** [comment]
Additionally, can you verify that the rest of VS is stable for you? It would help narrow down if this is a systemic VS install issue or something with my extension specifically.

**Comment by KiillThemAll** [comment]
Mooving back to Intellij rust)

Havent seen any problems yet with VS, but usually work in Qt Creator, therefore its hard to say.
unfortunately, there isnt appropriate plugins for creator(
everythings abandoned

**Comment by illera88** [comment]
Hi,

I'm having problems too. I followed the installation steps and everything went smooth. I have VS2107 Professional Preview on a Windows 10.
I've cloned https://github.com/bozaro/daemon-rs and opened the cloned `daemon-rs` folder in VS.
I've installed `rustup install nightly-x86_64-pc-windows-msvc`.

```
rustup.exe --version
rustup 1.9.0 (57fc3c087 2018-01-04)
```
I've set `rustup` in VS config:
<img width="466" alt="2018-01-14 13_16_10-options" src="https://user-images.githubusercontent.com/5193128/34915847-5c0530cc-f92d-11e7-9dcc-7d3614e563bc.png">

But don't see anything compiled when I open a `.rs` file. 

What other info can I provide so we can debug the problem?

Cheers

**Comment by dgriffen** [comment]
First verify that the rls is included in your current nightly build. You can check this with `rustup component list --toolchain nightly`. If you see `rls-preview` then the rls was included. If not I recommend setting your toolchain to `nightly-2017-12-08`. Also not that setting the path to rustup isn't necessary if rustup is already on your path.

Once you have that done please tell me which messages you see at the top when you open a .rs file. You should see messages to install the toolchain and install the rls components. If you don't see either of these make sure you opened a cargo project folder with the VS open folder feature.

If you still don't see any operation then I have to conclude that `System.ValueTuple` is broken for you as well. Which is completely ridiculous that this keeps happening since the language server extension also relies on it.

**Comment by illera88** [comment]
Hi,

First of all thank you for your help.
RLS is installed:

```
C:\Users\default3\Documents\code\daemon-rs>rustup component list --toolchain nightly
cargo-x86_64-pc-windows-msvc (default)
rls-preview-x86_64-pc-windows-msvc (installed)
rust-analysis-x86_64-pc-windows-msvc (installed)
rust-docs-x86_64-pc-windows-msvc (default)
rust-src (installed)
rust-std-aarch64-apple-ios
```

I've done this steps now:
`cargo new --bin "test"`
and from VS2017 Preview File -> Open Folder  and select the test file that has the `Cargo.toml` and `src` folder with the `main.rs` file. But nothing happens when I open the `.rs` file :(

What other info do you need? 

Cheers

**Comment by illera88** [comment]
Btw the first time that I opened a project it did showed something that I pressed "Install" but I don't see that anymore.

**Comment by dgriffen** [comment]
You won't see the install prompts the second time since installation has already happened. One last thing you could do is install the extra info bars version I posted earlier in the thread. That will give us more insight into where it's stopping. It's almost certainly the same thing though.

**Comment by dgriffen** [comment]
@illera88 would you be able to get me the exact preview version you have? You can find this in help > about

**Comment by dgriffen** [comment]
What do you mean you cannot select the main.rs file? I see the file open in your screenshot.

**Comment by illera88** [comment]
I saw a popup saying `"Requiered components are not installed"` And I clicked on `Install`

**Comment by illera88** [comment]
I can't set the `main.rs` as startup file so I see this:
<img width="351" alt="2018-01-18 20_18_39-test-test - microsoft visual studio preview" src="https://user-images.githubusercontent.com/5193128/35116747-cf521754-fc8c-11e7-920a-76659e0dbf2b.png">

What should I do to compile `main.rs`?

**Comment by dgriffen** [comment]
That would be because Visual Studio's open folder mode can't set source files as startup items. You can right click on main.rs in your Solution Explorer and select "Configure Tasks" This will allow you to create configurations for launching your Rust project. For more information on `launch.vs.json` refer to [this](https://blogs.msdn.microsoft.com/vcblog/2016/10/05/bring-your-c-codebase-to-visual-studio-with-open-folder/) page, specifically the "Building C++ Projects" and "Debugging C++ Binaries" sections. The same rules apply to Rust, just switch out the build commands for cargo.

**Comment by illera88** [comment]
Can you paste an example `launch.vs.json` I can use for rust to be used?

Thank you!

**Comment by dgriffen** [comment]
```json
{
  "version": "0.2.1",
  "tasks": [
    {
      "taskName": "task-main",
      "appliesTo": "src/main.rs",
      "type": "launch",
      "command": "cmd.exe",
      "args": [ "cargo", "run" ]
    }
  ]
}
```

If you have any more questions this should probably move to the [gitter](https://gitter.im/rls-vs2017/Lobby) page.

**Comment by streamholder** [comment]
I have the same problem.
Running VS 15.7.0 Preview 2.0, latest nightly (which includes rls), on Windows 10.

**Comment by l7l7** [comment]
Unfortunately the issue persists at least for me.

VS community 15.0.9 review 2.0
rustup toolchain list:  beta-i686-pc-windows-msvc (default)
rustc --version: 1.30.0-beta.7 (0ebb25088 2018-09-22)
Windows 7.

When I install the extension, configure the path to rustup and then open folder with a cargo project I get no prompt or message to install the RLS. Tried to install it manually (went w/o warning/error) but it changed nothing in the plugin, e.g. it doesn't seem to work nor does it spawn rls.exe process.

The super infobar debug solution you've provided doesn't seem to work with new Visual Studio Preview as it relies on language server plugin which is ditched now per my understanding and the language server protocol became build-in functionality

Later I'll try to run it on Win10 virtual machine to see if this issue is OS-specific but I think you should make the process of installing RLS user-invocable and more verbose Maybe you could kindly update the debug version of the plugin so it can be launched on newer VS versions.

**Comment by dgriffen** [comment]
@l7l7 do you happen to know what version of the .NET framework you have installed?

**Comment by l7l7** [comment]
@dgriffen 
I have .NET framework version 4.7.

**Comment by dgriffen** [comment]
I think I figured out the issue, the version of `System.ValueTuple` I was referencing was newer than the one shipped with VS so there was probably a load order bug happening there. To make this simpler to reason about I'm just going to remove my use of `ValueTuple`

**Comment by dgriffen** [comment]
This should be fixed in `0.4.5` with the removal of `ValueTuple`. Let me know if there are still issues.
